src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -2598,39 +2598,63 @@ struct
 
   let protocol_state =
     io_field "protocolState"
-      ~doc:"Get the current protocol state, optionally encoded in Base64"
+      ~doc:
+        "Get the protocol state for a given block, optionally encoded in Base64"
       ~typ:(non_null string)
       ~args:
         Arg.
-          [ arg "encoding" ~doc:"Encoding format (JSON or BASE64)"
+          [ arg "stateHash" ~doc:"The state hash of the desired block"
+              ~typ:string
+          ; arg "height"
+              ~doc:"The height of the desired block in the best chain" ~typ:int
+          ; arg "encoding" ~doc:"Encoding format (JSON or BASE64)"
               ~typ:
                 (enum "Encoding"
                    ~values:
                      [ enum_value "JSON" ~value:`JSON
                      ; enum_value "BASE64" ~value:`BASE64
                      ] )
           ]
-      ~resolve:(fun { ctx = mina; _ } () encoding_opt ->
-        match Mina_lib.best_tip mina with
-        | `Active best_tip ->
-            let protocol_state =
-              Transition_frontier.Breadcrumb.protocol_state best_tip
-            in
-            let encoded =
-              match encoding_opt with
-              | Some `BASE64 ->
-                  Bin_prot.Writer.to_string
-                    Mina_state.Protocol_state.Value.Stable.V2.bin_t.writer
-                    protocol_state
-                  |> Base64.encode_exn
-              | Some `JSON | None ->
-                  (* Default to JSON if no encoding is specified *)
-                  Mina_state.Protocol_state.value_to_yojson protocol_state
-                  |> Yojson.Safe.to_string
-            in
-            return (Ok encoded)
-        | `Bootstrapping ->
-            return (Error "Node is bootstrapping") )
+      ~resolve:(fun { ctx = mina; _ } () state_hash_base58_opt height_opt
+                    encoding_opt ->
+        let open Deferred.Result.Let_syntax in
+        let%bind breadcrumb =
```

Comment:
```suggestion
        let%map breadcrumb =
```

NIT, but this would allow to remove `Deferred.Result.return` below and make the whole code a bit less multilayered.

Reply:
done! https://github.com/MinaProtocol/mina/pull/15893/commits/882f94e5cc0373a9367af832f13d5e7953c4f550